File: ProgrammingChallenges#Book/Chapter 1/06.Interpreter/solutions/Interpreter.java
import java.io.BufferedReader;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

/**
 *
 * @author XergioAleX
 */
public class Interpreter {

    public static class Computer {
        int address = 0;
        int[] registers = new int[10];
        String[] RAM = new String[1000];

        public Computer(){
            for (int i = 0; i < 1000; i++) {
                this.RAM[i] = "000";
            }
        }
        
        public int runProgram() {
            int countInstructions = 1;
            while (this.RAM[this.address].charAt(0) != '1') {
                countInstructions++;
                switch (this.RAM[this.address].charAt(0)) {
                    case '0':
                        this.instruction0(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '2':
                        this.instruction2(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '3':
                        this.instruction3(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '4':
                        this.instruction4(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '5':
                        this.instruction5(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '6':
                        this.instruction6(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '7':
                        this.instruction7(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '8':
                        this.instruction8(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                    case '9':
                        this.instruction9(this.RAM[this.address].charAt(1)-48, this.RAM[this.address].charAt(2)-48);
                        break;
                }
                
                this.address++;
            }
            
            return countInstructions;
        }

        public void instruction0(int d, int s){
            if (this.registers[s] != 0) {
                this.address = this.registers[d]-1;
            }
        }
        
        public void instruction2(int d, int n){
            this.registers[d] = n;
            this.registers[d] %= 1000;
        }
        
        public void instruction3(int d, int n){
            this.registers[d] += n;
            this.registers[d] %= 1000;
        }
        
        public void instruction4(int d, int n){
            this.registers[d] *= n;
            this.registers[d] %= 1000;
        }
        
        public void instruction5(int d, int s){
            this.registers[d] = this.registers[s];
            this.registers[d] %= 1000;
        }
        
        public void instruction6(int d, int s){
            this.registers[d] += this.registers[s];
            this.registers[d] %= 1000;
        }
        
        public void instruction7(int d, int s){
            this.registers[d] *= this.registers[s];
            this.registers[d] %= 1000;
        }
        
        public void instruction8(int d, int a){
            this.registers[d] = Integer.parseInt(this.RAM[this.registers[a]]);
            this.registers[d] %= 1000;
        }
        
        public void instruction9(int s, int a){
            this.RAM[this.registers[a]] = Integer.toString(this.registers[s]);
        }

    }

    public static void main(String[] XergioAleX) throws FileNotFoundException, IOException{
        BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer token;
        String inputLine;

        int cases = Integer.parseInt(reader.readLine());
        reader.readLine();
        
        for (int i = 0; i < cases; i++) {
            Computer pc = new Computer();
            
            int index = 0;
            while ((inputLine = reader.readLine()) != null && !inputLine.equals("")) {
                pc.RAM[index] = inputLine;
                index++;
            }
            
            System.out.println(pc.runProgram());
            
            if (i != cases-1) {
                System.out.println("");
            }
        }
        
    }
}

File: ProgrammingChallenges#Book/Chapter 1/06.Interpreter/solutions/Interpreter.py
from sys import stdin


class Computer:
    def __init__(self):
        self.address = 0
        self.registers = [0 for x in range(10)]
        self.RAM  = ["000" for x in range(1000)]

        
    def runProgram(self):
        def case0():
            self.instruction0(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case2():
            self.instruction2(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case3():
            self.instruction3(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case4():
            self.instruction4(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case5():
            self.instruction5(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case6():
            self.instruction6(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case7():
            self.instruction7(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case8():
            self.instruction8(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))

        def case9():
            self.instruction9(int(self.RAM[self.address][1]), int(self.RAM[self.address][2]))
            
        switch = {
            '0': case0,
            '2': case2,
            '3': case3,
            '4': case4,
            '5': case5,
            '6': case6,
            '7': case7,
            '8': case8,
            '9': case9
        }

        countInstructions = 1
        while (self.RAM[self.address][0] != '1'):
            countInstructions += 1

            if (self.RAM[self.address][0] in switch.keys()):
                switch[self.RAM[self.address][0]]()
            self.address += 1
        
        return countInstructions



    def instruction0(self, d, s):
        if (self.registers[s] != 0):
            self.address = self.registers[d]-1
    
    def instruction2(self, d, n):
        self.registers[d] = n
        self.registers[d] %= 1000
    
    def instruction3(self, d, n):
        self.registers[d] += n
        self.registers[d] %= 1000
    
    def instruction4(self, d, n):
        self.registers[d] *= n
        self.registers[d] %= 1000
    
    def instruction5(self, d, s):
        self.registers[d] = self.registers[s]
        self.registers[d] %= 1000
    
    def instruction6(self, d, s):
        self.registers[d] += self.registers[s]
        self.registers[d] %= 1000
    
    def instruction7(self, d, s):
        self.registers[d] *= self.registers[s]
        self.registers[d] %= 1000
    
    def instruction8(self, d, a):
        self.registers[d] = int(self.RAM[self.registers[a]])
        self.registers[d] %= 1000
    
    def instruction9(self, s, a):
        self.RAM[self.registers[a]] = str(self.registers[s])



# Main program
def main():    
    cases = int(stdin.readline())
    stdin.readline()
    
    for i in range(0, cases):
        pc = Computer()
        
        index = 0;
        for inputLine in stdin:
            if (inputLine == "" or inputLine == "\n"):
                break

            pc.RAM[index] = inputLine
            index += 1
        
        print(pc.runProgram())
        
        if (i != cases-1):
            print()
        

main()
